madara/crates/client/rpc/src/versions/user/v0_7_1/methods/read/syncing.rs

```diff
@@ -15,32 +15,112 @@ use crate::Starknet;
+        // We need to fetch the starting block hash from the backend
+        let Some(starting_block_info) = starknet
+            .backend
+            .get_block_info(&BlockId::Number(starting_block_num))
+            .or_internal_server_error("Error getting starting block")?
+        else {
+            // If we can't get the starting block info, use a default hash
+            // This is a fallback and shouldn't normally happen
+            tracing::warn!("Could not get starting block info for block {}", starting_block_num);
+            return Ok(SyncingStatus::NotSyncing);
+        };
+
+        let starting_block_info =
+            starting_block_info.as_nonpending().ok_or_internal_server_error("Starting block cannot be pending")?;
+
+        starting_block_info.block_hash
+    } else {
+        // Use the hash from sync status
+        sync_status.starting_block_hash
+    };
+
+    // Get highest block info from sync status
+    let highest_block_num = sync_status.highest_block_num;
+    let highest_block_hash = sync_status.highest_block_hash;
+
+    // If highest block number is 0 or less than current, we're not syncing
+    if highest_block_num == 0 || highest_block_num - 6 <= current_block_num {
+        return Ok(SyncingStatus::NotSyncing);
+    }
+
     Ok(SyncingStatus::Syncing(SyncStatus {
         starting_block_num,
         starting_block_hash,
-        highest_block_num: current_block_num, // TODO(merge): is this correct?
-        highest_block_hash: current_block_hash,
+        highest_block_num,
+        highest_block_hash,
         current_block_num,
         current_block_hash,
     }))
 }
+
+// Helper function to get sync status with exponential backoff retry
+async fn get_sync_status_with_retry(starknet: &Starknet) -> Result<mp_rpc::SyncStatus, anyhow::Error> {
+    use std::time::Duration;
+    use tokio::time;
+
+    const MAX_RETRIES: u32 = 3;
+    const INITIAL_TIMEOUT_MS: u64 = 1000;
+
+    let mut retries = 0;
+    let mut timeout_ms = INITIAL_TIMEOUT_MS;
+
+    loop {
+        match time::timeout(Duration::from_millis(timeout_ms), starknet.sync_status()).await {
+            // Successful response within timeout
+            Ok(Ok(status)) => return Ok(status),
+
+            // Timeout occurred
+            Err(_) => {
+                tracing::warn!("Timeout getting sync status after {}ms", timeout_ms);
+            }
+
+            // Error occurred within timeout
+            Ok(Err(err)) => {
+                tracing::warn!("Error getting sync status: {}", err);
+            }
+        }
+
+        // Check if we've reached the maximum number of retries
+        retries += 1;
+        if retries >= MAX_RETRIES {
+            return Err(anyhow::anyhow!("Failed to get sync status after {} retries", MAX_RETRIES));
+        }
+
+        // Exponential backoff
+        timeout_ms *= 2;
+        tracing::debug!("Retrying sync status (attempt {}/{}), timeout: {}ms", retries + 1, MAX_RETRIES, timeout_ms);
```

Comment:
Shouldn't be like:
```suggestion
        tracing::debug!("Retrying sync status (attempt {}/{}), timeout: {}ms", retries, MAX_RETRIES, timeout_ms);
```
since you're already incrementing `retries` above